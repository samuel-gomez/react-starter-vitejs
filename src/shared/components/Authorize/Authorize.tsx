import { UserContext } from 'App/UserProvider';
import { useContext, type ReactNode } from 'react';

type TAuthorize = {
  authorized?: string[];
  children: ReactNode;
  UserContextObj?: typeof UserContext;
};

const Authorize = ({ children, authorized, UserContextObj = UserContext }: TAuthorize) => {
  const { authRole } = useContext(UserContextObj);
  return authorized !== undefined && !authorized.includes(authRole) ? null : <>{children}</>;
};

export default Authorize;
