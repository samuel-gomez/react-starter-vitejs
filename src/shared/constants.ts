import packageJson from '../../package.json';

/** ***********************************************************************
 * FORM CONSTANTS
 ************************************************************************ */
export const MSG_REQUIRED = 'Le champ est obligatoire';

export const MSG_LENGTH_3 = 'Le champ doit faire au moins 3 caractères';
export const MSG_FORMAT = 'Format invalide';
export const MSG_ERROR_FORMAT = 'Format de fichier incorrect';

export const DEFAULT_OPTION_LABEL = '- Sélectionner -';
export const DEFAULT_OPTION = {
  value: '',
  label: DEFAULT_OPTION_LABEL,
};

// patterns
export const patternEmail = /^[\w-+]+@([\w-+]+\.)+[\w-]{2,}$/;
export const patternPhoneNumber = /^[+0][\d ]*$/;
export const patternZipCode = /^\d{5}$/;
export const patternSiret = /^\d{14}$/;
export const patternSiren = /^\d{9}$/;
export const patternSiretSiren = /^\d{9,14}$/;

export const phoneNumberRule = {
  required: true,
  minLength: 4,
  maxLength: 20,
  pattern: patternPhoneNumber,
};

export const emailRule = {
  required: true,
  minLength: 6,
  maxLength: 60,
  pattern: patternEmail,
};

/** ***********************************************************************
 * API STATUS CONSTANTS
 ************************************************************************ */
export const STATUS_HTTP = {
  SUCCESS: 200,
  CREATED: 201,
  BAD_REQUEST: 400,
  UNAUTHORIZE: 401,
  FORBIDDEN: 403,
  NOTFOUND: 404,
  SERVER_ERROR: 500,
};

export const STATUS_API = {
  SUCCESS: STATUS_HTTP.SUCCESS,
  CREATED: STATUS_HTTP.CREATED,
  WARNING: 40,
  ERROR: 50,
};

/** ***********************************************************************
 * API URL
 ************************************************************************ */
export const API_URL = {
  BASE: 'base',
  GITHUB: 'github',
};

export const STATUS_HTTP_MESSAGES = {
  [STATUS_HTTP.SUCCESS]: 'Succès: La requête a réussi',
  [STATUS_HTTP.CREATED]: 'Succès: Création réussie',
  [STATUS_HTTP.BAD_REQUEST]: 'Erreur: La syntaxe de la requête est erronée',
  [STATUS_HTTP.UNAUTHORIZE]: 'Erreur: Une authentification est nécessaire pour accéder à la ressource',
  [STATUS_HTTP.FORBIDDEN]: 'Erreur: Accès non autorisé',
  [STATUS_HTTP.NOTFOUND]: 'Erreur: Elément non trouvé',
  [STATUS_HTTP.SERVER_ERROR]: 'Erreur: Problème technique ! Contacter votre support',
};

/** ***********************************************************************
 * USER CONSTANTS
 ************************************************************************ */

export const ADMIN = 'Admin';
export const READER = 'Reader';
export const WRITER = 'Writer';
export const USER = 'User';
export const ALL = '';

export const PROFILS = [ADMIN, READER, WRITER, USER, ALL];

/** ***********************************************************************
 * COMMONS CONSTANTS
 ************************************************************************ */

export const MODIFIER_CLASS = {
  success: 'success',
  disabled: 'disabled',
};

/** ***********************************************************************
 * VERSION CONSTANT
 ************************************************************************ */

const VERSION_TOOLKIT = 'master';

/** ***********************************************************************
 * DEMO CONSTANTS
 ************************************************************************ */

export const STORYBOOK = 'https://axafrance.github.io/react-toolkit/latest/storybook/?path=/story/';
export const DESIGN_SYSTEM = 'https://axafrance.github.io/design-system/';
export const GITHUB = `https://github.com/AxaGuilDEv/react-toolkit/tree/v${packageJson.dependencies['@axa-fr/react-toolkit-all']}/packages/`;
export const GITHUB_API = `AxaGuilDEv/react-toolkit/${VERSION_TOOLKIT}/packages/`;
