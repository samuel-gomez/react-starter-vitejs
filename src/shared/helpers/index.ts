export { formatDate, isValidDate, setDate } from './formatDate';
export { default as isEmptyOrNull, isEmptyArray, isEmptyObject } from './isEmptyOrNull';
export { default as mergeObj } from './mergeObj';
export { default as setAnomalyEmptyItems } from './setAnomalyEmptyItems';
export { default as setConfirmClassModifier } from './setConfirmClassModifier';
export { default as emptyFunction } from './emptyFunction';
export { default as setRequiredFieldRule, setRequiredMessage } from './setRequiredFieldRule';
