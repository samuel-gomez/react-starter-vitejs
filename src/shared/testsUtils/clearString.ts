const cleanString = (str?: string) => str?.replace(/\s/g, '').replace(/\n/g, '').replace(/\t/g, '').trim();

export default cleanString;
