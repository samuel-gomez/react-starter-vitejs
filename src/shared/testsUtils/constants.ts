const MOCK_API_URL = { base: 'https://react-starter-api.vercel.app/api/', github: 'https://raw.githubusercontent.com/' };

export default MOCK_API_URL;
