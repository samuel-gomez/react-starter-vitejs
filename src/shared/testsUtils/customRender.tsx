import { vi } from 'vitest';
import { render, RenderOptions } from '@testing-library/react';
import { StaticRouter } from 'react-router-dom/server';
import FetchProvider from 'App/FetchProvider';
import UserProvider from 'App/UserProvider';
import { ReactElement, ReactNode } from 'react';
import MOCK_API_URL from './constants';

type TMockProvider = {
  [x: string]: Record<string, unknown | number | string> | string;
};

const MockProviders =
  ({ role = '', ...testMock }: TMockProvider) =>
  ({ children }: { children: ReactNode }) => {
    const useOidcAccessTokenFn = vi.fn().mockReturnValue({ accessToken: 'accessToken' });
    const useOidcUserFn = vi.fn().mockReturnValueOnce({ oidcUser: { member_of: [`CN=${role}`], name: 'Samuel Gomez' } });
    return (
      <UserProvider useOidcUserFn={useOidcUserFn}>
        <FetchProvider
          apiUrl={MOCK_API_URL}
          fetchConfig={{
            headers: { testMock: JSON.stringify(testMock) },
          }}
          useOidcAccessTokenFn={useOidcAccessTokenFn}
        >
          <StaticRouter location="">{children}</StaticRouter>
        </FetchProvider>
      </UserProvider>
    );
  };

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, 'wrapper'>, testMock: TMockProvider = { role: '' }) =>
  render(ui, { wrapper: MockProviders(testMock), ...options });

// re-export everything
export * from '@testing-library/react';
export { default as userEvent } from '@testing-library/user-event';

// override render method
export { customRender as render };
