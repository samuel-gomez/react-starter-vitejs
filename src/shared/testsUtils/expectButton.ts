import { screen } from '@testing-library/dom';

type TexpectButton = {
  name: string;
  isQueryByRole?: boolean;
  beDisabled?: boolean;
  beInDoc?: boolean;
};

const expectButton = ({ isQueryByRole = false, name, beDisabled = true, beInDoc = true }: TexpectButton) => {
  const button = isQueryByRole ? screen.queryByRole('button', { name: RegExp(name) }) : screen.getByRole('button', { name: RegExp(name) });
  if (beInDoc) {
    expect(button).toBeInTheDocument();
    if (beDisabled) {
      expect(button).toBeDisabled();
    } else {
      expect(button).not.toBeDisabled();
    }
  } else {
    expect(button).not.toBeInTheDocument();
  }
};

export default expectButton;
