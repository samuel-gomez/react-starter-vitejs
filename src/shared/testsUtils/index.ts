export * from './customRender';
export { default as emptyFunction } from '../helpers/emptyFunction';
export { default as MOCK_API_URL } from './constants';
export { default as renderWithContainer } from './renderWithContainer';
