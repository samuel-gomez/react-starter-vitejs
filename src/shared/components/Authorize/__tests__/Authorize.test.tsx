import { render, screen } from 'shared/testsUtils';
import Authorize from '../Authorize';

describe('<Authorize/>', () => {
  const text = 'protected content';

  it.each`
    childrenText | authorized   | role
    ${text}      | ${['admin']} | ${undefined}
    ${text}      | ${['admin']} | ${'Admin'}
  `('Should render null, childrenText: $childrenText, authorized: $authorized, role: $role', async ({ childrenText, authorized, role }) => {
    render(
      <Authorize authorized={authorized}>
        <p>{childrenText}</p>
      </Authorize>,
      {},
      { role },
    );

    expect(screen.queryByText(childrenText)).toBeNull();
  });

  it.each`
    childrenText | authorized           | role
    ${text}      | ${undefined}         | ${undefined}
    ${text}      | ${['Admin', 'User']} | ${'Admin'}
  `('Should render children, childrenText: $childrenText, authorized: $authorized, role: $role', async ({ childrenText, authorized, role }) => {
    render(
      <Authorize authorized={authorized}>
        <p>{childrenText}</p>
      </Authorize>,
      {},
      { role },
    );

    expect(screen.getByText(childrenText)).toBeInTheDocument();
  });
});
