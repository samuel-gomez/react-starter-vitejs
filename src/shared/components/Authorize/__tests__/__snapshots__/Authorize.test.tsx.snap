// Vitest Snapshot v1

exports[`<Authorize/> > Render <Authorize />  1`] = `<DocumentFragment />`;

exports[`<Authorize/> > Render <Authorize /> with child 1`] = `
<DocumentFragment>
  <p>
    child
  </p>
</DocumentFragment>
`;

exports[`<Authorize/> > Render child when user role is authorized 1`] = `
<DocumentFragment>
  <p>
    child
  </p>
</DocumentFragment>
`;

exports[`<Authorize/> > Render null with child and unauthorize user role 1`] = `<DocumentFragment />`;
